### PR TITLE
Feature: Record tx expenses

### PR DESCRIPTION
**Pull request by dpatsora**
- [x] rename event from confirm to wait
- [x] fee_processing state should be changed
- [x] update peatio core version
- [x] rebase PR on master branch
- [ ] issue with DB migration, how we should deal with deposits in processing/fee_processing state
- [x] refactor code in blockchain_service
- [x] split process_existing_deposits into process_fee_tx and process_collection_tx
- [ ] fireblocks/bitgo plugins should be compatible (logic in webhook_helpers)
- [ ] documentation
- [ ] rspecs

### CODE REVIEW

File: app/models/deposit.rb [hunk 1]
```diff
@@ -45,8 +45,10 @@ class Deposit < ApplicationRecord
     state :aml_suspicious
     state :processing
     state :skipped
+    state :collecting
```
Comment (dpatsora) [inline app/models/deposit.rb:48]: maybe better `collection`?

File: app/models/transaction.rb [hunk 2]
```diff
@@ -1,47 +1,90 @@
...
+
+  def record_expenses!
+    return unless fee?
+
+    Operations::Expense.create!({
+                                  code: 402,
+                                  currency_id: fee_currency_id,
+                                  reference_id: reference_id,
+                                  reference_type: reference_type,
+                                  debit: fee,
```
Comment (dpatsora) [inline app/models/transaction.rb]: is it right?

File: app/services/blockchain_service.rb [hunk 3]
```diff
@@ -70,11 +73,55 @@ def update_height(block_number)
 
     # NOTE: We use update_column to not change updated_at timestamp
     # because we use it for detecting blockchain configuration changes see Workers::Daemon::Blockchain#run.
-    blockchain.update_column(:height, block_number) if latest_block_number - block_number >= blockchain.min_confirmations
+    if latest_block_number - block_number >= blockchain.min_confirmations
+      blockchain.update_column(:height, block_number)
+    end
```
Comment (dpatsora) [inline app/services/blockchain_service.rb:77]: Rubocop

File: app/services/blockchain_service.rb [hunk 4]
```diff
@@ -70,11 +73,55 @@ def update_height(block_number)
...
+
+  # Deposit in state fee_processing
+  # There is no Transaction yet
+  # no actions
+
+  # Deposit in state collecting
+  # check tx state
+  # if succeed change state to collected and change state of db_tx to succeed
+  def process_pending_deposit_txs(deposit_txs)
+    # add select by blockchain_key
```
Comment (dpatsora) [inline app/services/blockchain_service.rb]: What is the point of selecting by blockchain_key?

File: app/services/blockchain_service.rb [hunk 5]
```diff
@@ -70,11 +73,55 @@ def update_height(block_number)
...
+    db_txs = Transaction.where(txid: deposit_txs.map(&:hash), reference_type: Deposit, status: 'pending')
+
+    db_txs.each do |db_tx|
+      deposit_tx = deposit_txs.find { |tx| tx if db_tx.txid == tx.hash }
+      deposit = db_txs.reference
+
+      tx = adapter.fetch_transaction(tx) if @adapter.respond_to?(:fetch_transaction) && transaction.status.pending?
```
Comment (dpatsora) [inline app/services/blockchain_service.rb]: Can we be sure, that fees always will be returned here?

File: app/services/blockchain_service.rb [hunk 6]
```diff
@@ -70,11 +73,55 @@ def update_height(block_number)
...
+      if tx.status == 'success'
+        deposit_tx.confirm!
+        deposit.fee_process! if deposit.fee_collecting?
+        deposit.dispatch! if deposit.collecting?
+      else
+        deposit_tx.fail!
+        deposit.err! 'Fee collection transaction failed'
+      end
+
+      next
```
Comment (dpatsora) [inline app/services/blockchain_service.rb]: remove

File: app/services/blockchain_service.rb [hunk 7]
```diff
@@ -96,17 +143,27 @@ def update_or_create_deposit(transaction)
       return
     end
 
+    # In which cases????
     # Fetch transaction from a blockchain that has `pending` status.
-    transaction = adapter.fetch_transaction(transaction) if @adapter.respond_to?(:fetch_transaction) && transaction.status.pending?
+    if @adapter.respond_to?(:fetch_transaction) && transaction.status.pending?
+      transaction = adapter.fetch_transaction(transaction)
+    end
```
Comment (dpatsora) [inline app/services/blockchain_service.rb]: Maybe we can remove this block?

File: lib/peatio/ethereum/blockchain_abstract.rb [hunk 8]
```diff
@@ -193,14 +195,15 @@ def build_transactions(tx_hash)
...
-          status:         transaction_status(block_txn)
+          hash:             normalize_txid(block_txn.fetch('hash')),
+          amount:           convert_from_base_unit(block_txn.fetch('value').hex, @eth),
+          from_addresses:   [normalize_address(block_txn['from'])],
+          to_address:       normalize_address(block_txn['to']),
+          txout:            block_txn.fetch('transactionIndex').to_i(16),
+          block_number:     block_txn.fetch('blockNumber').to_i(16),
+          currency_id:      @eth.fetch(:id),
+          fee_currency_id:  @eth.fetch(:id),
+          status:           transaction_status(block_txn)
```
Comment (dpatsora) [inline lib/peatio/ethereum/blockchain_abstract.rb:213]: add fee

File: app/models/transaction.rb (hunk 2 above)
Comment (mnaichuk) [reply app/models/transaction.rb]: should be credit

File: app/services/blockchain_service.rb [hunk 9]
```diff
@@ -70,14 +72,40 @@ def update_height(block_number)
...
   private
-
-  def filter_deposits(block)
+  # Filters deposit, fee and deposit_collection txs
+  # This method is so complicated to reduce amount of database interactions during deposit processing
+  def filter_deposit_txs(block)
+    # Filter transaction source/destination addresses
     addresses = PaymentAddress.where(wallet: Wallet.deposit.with_currency(@currencies.codes), address: block.transactions.map(&:to_address)).pluck(:address)
-    block.select { |transaction| transaction.to_address.in?(addresses) }
+    Wallet.hot.with_currency(@currencies.codes).pluck(:address).map do |addr|
```
Comment (mnaichuk) [inline app/services/blockchain_service.rb]: Wallet.withdraw

File: app/services/blockchain_service.rb [hunk 10]
```diff
@@ -86,6 +114,44 @@ def filter_withdrawals(block)
...
+      next unless block_deposit_tx
+
+      deposit = db_txs.reference
+      next unless deposit.fee_collecting? || deposit.collecting?
+
+      tx = adapter.fetch_transaction(tx) if @adapter.respond_to?(:fetch_transaction) && transaction.status.pending?
+      next unless tx.status.success?
+      deposit_tx.fee = tx.fee
+
+      if tx.status == 'success'
```
Comment (mnaichuk) [inline app/services/blockchain_service.rb]: check also all transactions belongs_to this deposit

File: app/models/transaction.rb [hunk 11]
```diff
@@ -1,47 +1,92 @@
...
+    end
+
+    event :fail do
+      transitions from: :pending, to: :failed
+      after do
+        record_expenses!
+      end
+    end
+
+    event :reject do
```
Comment (dpatsora) [inline app/models/transaction.rb:38]: Not sure if we need it

File: app/services/blockchain_service.rb [hunk 12]
```diff
@@ -70,14 +72,42 @@ def update_height(block_number)
...
   private
-
-  def filter_deposits(block)
+  # Filters deposit, fee and deposit_collection txs
+  # This method is so complicated to reduce amount of database interactions during deposit processing
+  def filter_deposit_txs(block)
+    # Filter transaction source/destination addresses
     addresses = PaymentAddress.where(wallet: Wallet.deposit.with_currency(@currencies.codes), address: block.transactions.map(&:to_address)).pluck(:address)
-    block.select { |transaction| transaction.to_address.in?(addresses) }
+    # TODO: add all withdraw wallets
```
Comment (dpatsora) [inline app/services/blockchain_service.rb]: no need

File: app/services/blockchain_service.rb [hunk 13]
```diff
@@ -96,22 +188,21 @@ def update_or_create_deposit(transaction)
       return
     end
 
+    # In which cases????
```
Comment (dpatsora) [inline app/services/blockchain_service.rb]: remove

File: lib/peatio/ethereum/blockchain_abstract.rb [hunk 14]
```diff
@@ -1,6 +1,9 @@
+# frozen_string_literal: true
+
+require 'pry-byebug'
```
Comment (dpatsora) [inline lib/peatio/ethereum/blockchain_abstract.rb:2]: remove

File: lib/peatio/ethereum/blockchain_abstract.rb [hunk 15]
```diff
@@ -225,27 +231,30 @@ def build_erc20_transactions(txn_receipt)
...
+                             txout: log['logIndex'].to_i(16),
+                             block_number: txn_receipt.fetch('blockNumber').to_i(16),
+                             currency_id: currency.fetch(:id),
+                             # fee_currency_id:    @eth.fetch(:id),
+                             status: transaction_status(txn_receipt) }
         end
       end
     end
 
+    # What is the point?
```
Comment (dpatsora) [inline lib/peatio/ethereum/blockchain_abstract.rb]: remove

File: app/services/blockchain_service.rb [hunk 16]
```diff
@@ -86,6 +116,68 @@ def filter_withdrawals(block)
...
+
+      block_tx = adapter.fetch_transaction(block_tx) if @adapter.respond_to?(:fetch_transaction) && ( block_tx.fee.blank? || block_tx.status.pending? )
+
+      db_tx.fee = block_tx.fee
+      # db_tx.fee_currency_id = block_tx.fee_currency_id
+      db_tx.block_number = block_tx.block_number
+      db_tx.save!
+
+      # BSC can return success only after fetch transaction
+      if block_tx.status == 'success'
```
Comment (dpatsora) [inline app/services/blockchain_service.rb]: block_tx.success?

**Comment by martianatwork** [comment]
Finally 🎉🎉

File: app/services/blockchain_service.rb [hunk 17]
```diff
@@ -86,6 +116,68 @@ def filter_withdrawals(block)
...
+  # if succeed change state to collected and change state of db_tx to succeed
+  def process_pending_deposit_txs(block_txs, db_txs)
+    db_txs.each do |db_tx|
+      next unless db_tx.pending?
+
+      block_tx = block_txs.find { |tx| tx if db_tx.txid == tx.hash }
+      next unless block_tx
+
+      deposit = db_tx.reference
+      next unless deposit.fee_collecting? || deposit.collecting?
```
Comment (chumaknadya) [inline app/services/blockchain_service.rb:142]: fee_processing

File: app/services/blockchain_service.rb [hunk 18]
```diff
@@ -86,6 +116,68 @@ def filter_withdrawals(block)
...
+      # db_tx.fee_currency_id = block_tx.fee_currency_id
+      db_tx.block_number = block_tx.block_number
+      db_tx.save!
+
+      # BSC can return success only after fetch transaction
+      if block_tx.success?
+        db_tx.confirm!
+
+        if deposit.fee_collecting? && db_tx.kind == 'tx_prebuild'
+          deposit.fee_process! if deposit.fee_collecting?
```
Comment (chumaknadya) [inline app/services/blockchain_service.rb:156]: deposit.fee_process!

**Pull request closed without merging**